pkg/dhcp6client/client.go

```diff
@@ -0,0 +1,496 @@
+//  ...
+//  iana, packet, err := c.RapidSolicit()
+//  ...
+//  // iana now contains the IP assigned in the IAAddr option.
+//
+//
+// Example selecting which advertising server to request from:
+//
+//   c, err := dhcp6client.New(iface)
+//   ...
+//   ctx, cancel := context.WithTimeout(context.Background(), 5 * time.Second)
+//   defer cancel()
+//
+//   ads, err := c.Solicit(ctx)
+//   ...
+//   // Selecting the advertisement of server 3.
+//   request, err := dhcp6client.RequestIANAFrom(ads[2])
+//   ...
+//   iana, packet, err := c.RequestOne(request)
+//   ...
+//   // iana now contains the IP assigned in the IAAddr option.
+type Client struct {
+	// The interface to send requests on.
+	iface netlink.Link
+
+	// Packet socket to send on.
+	conn net.PacketConn
+
+	// Max number of attempts to multicast DHCPv6 solicits.
+	// -1 means infinity.
+	retry int
+
+	// Timeout for each Solicit try.
+	timeout time.Duration
+}
+
+// New returns a new DHCPv6 client based on the given parameters.
+func New(iface netlink.Link, opts ...ClientOpt) (*Client, error) {
+	haddr := iface.Attrs().HardwareAddr
+	ip, err := eui64.ParseMAC(net.ParseIP("fe80::"), haddr)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := net.ListenUDP("udp6", &net.UDPAddr{
+		IP:   ip,
+		Port: ClientPort,
+		Zone: iface.Attrs().Name,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Client{
+		iface:   iface,
+		conn:    conn,
+		timeout: 10 * time.Second,
+		retry:   3,
+	}
+
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+	return c, nil
+}
+
+// ClientOpt is a function that configures the client.
+type ClientOpt func(*Client) error
+
+// WithTimeout configures the retransmission timeout.
+//
+// Default is 10 seconds.
+//
+// TODO(hugelgupf): Check RFC for retransmission behavior.
+func WithTimeout(d time.Duration) ClientOpt {
+	return func(c *Client) error {
+		c.timeout = d
+		return nil
+	}
+}
+
+// WithRetry configures the retransmission counts.
+//
+// Default is 3.
+//
+// TODO(hugelgupf): Check RFC for retransmission behavior.
+func WithRetry(retry int) ClientOpt {
+	return func(c *Client) error {
+		c.retry = retry
+		return nil
+	}
+}
+
+// RapidSolicit solicits one non-temporary address assignment by multicasting a
+// DHCPv6 solicitation message with the rapid commit option.
+//
+// RapidSolicit returns the first valid, suitable response by any remote server.
+func (c *Client) RapidSolicit() (*dhcp6opts.IANA, *dhcp6.Packet, error) {
+	solicit, err := NewRapidSolicit(c.iface.Attrs().HardwareAddr)
+	if err != nil {
+		return nil, nil, err
+	}
+	return c.RequestOne(solicit)
+}
+
+// RequestOne multicasts the `request` and returns the first matching IANA and
+// its associated Packet returned by any server.
+func (c *Client) RequestOne(request *dhcp6.Packet) (*dhcp6opts.IANA, *dhcp6.Packet, error) {
+	ianas, pkt, err := c.Request(request)
+	if err != nil {
+		return nil, nil, err
+	}
+	if len(ianas) != 1 {
```

Comment:
technically no - as per RFC 3315 (chap. 22 par. 4), "A DHCP message may contain multiple IA_NA options.". Because of this I prefer to leave the current behaviour.